### ISSUE TITLE
Retain GPO (PIN) status on re-configuration

### ISSUE THREAD

**Issue by mschuetterle** [eblot/pyftdi#218]
Hi,

We are using PyFtdi with a dual FTDI (FT2232) chip by connecting port A in MPSSE mode:
```
port = GpioMpsseController()
/# Configure Bit13 as Output
port.configure('ftdi://ftdi:ft2232/1', direction=0x2000, frequency=10e6)
```

We have seen that after toggling Bit13 and closing the port `port.close()` the value got retained.
**But when re-running the SW and again configuring the port, the Bit13 got a reset!**

Even using `direction=0` to read Bit13 first, does not change the behavior. So, my assumption, this is an issue on the configuration of the Port into MPSSE mode.

Also the `initial` value does not help, as we just want to retain the previous value of Bit13 - which is unknown.

Is there any idea or other way to resolve this?

**Comment by eblot** [comment]
I would say this is by design: whenever you close the port, you release the hardware resources, and revert them to the idle state. I guess you could hack into the code to prevent PyFtdi from performing a port reset on release.

Whenever the port is open, the interface is reset - because there is no other way to start from a known state - so I'm afraid re-opening always means resetting.

Why do you close the port if you need to keep using the device BTW?

**Comment by mschuetterle** [comment]
Thank you for the feedback.

Just want to correct a misunderstanding:
The hardware resources are not been reverted to idle state when closing the port, but when opening (re-configuring) the next time. This was something unexpected to me.

Your last question is perfectly the right one. Let me help to clarify:
- Our SW could be considered a Daemon running in the background.
- The PIN13 is a RESET signal for other HW and will release them when our SW is started.
- Closing the port is just part of cleanup in our SW for example when shutting down the system.

Now, why I'm asking?
We are doing some test which require to frequently restart our SW (Daemon) and this is causing each time a reset cycle on the connected HW.
So, I tried to find a way that the PIN state from previous run will not be changed.

Thank you for your support

**Comment by eblot** [comment]
Ok I got it.

A heavy duty solution would be to fork a daemon to manage FTDI, so that it stays alive even if the SW daemon is reset.

Unfortunately there is no way to know the current state of the FTDI HW, so not resetting it whenever the port is open would lead to unpredictable behaviour, but I guess you can hack you own copy of PyFtdi to discard all reset class and see if it helps, but I'm afraid it gonna lead to some weird behaviour one day or another.

**Comment by mschuetterle** [comment]
Okay, again thank you for the feedback!

**Comment by eblot** [comment]
There is now an option to preserve the output state of GPIO pin when the port is closed, see #219.

However, it does not address this initial requirement: the port will still be re-initialized on open.

**Comment by woodrowbarlow** [comment]
hello eblot! thanks for this great package.

is it possible to add something like the `freeze` parameter for opening ports as well? i'd like to be able to check the pin state without keeping the device connection open for the full lifespan of the application.

here's my example use-case:

```python
import contextlib
import enum
from pyftdi.ftdi import Ftdi


class MuxMode(enum.IntEnum):
    TARGET = 0
    HOST = 1


class SDWireMux:
    def __init__(self, url: str):
        self.url = url

    @contextlib.contextmanager
    def _ftdi_context(self):
        ctx = Ftdi()
        ctx.open_from_url(self.url)
        try:
            yield ctx
        finally:
            ctx.close(freeze=True)

    @property
    def mode(self) -> MuxMode:
        with self._ftdi_context() as ctx:
            pins = ctx.read_pins()
            return MuxMode(pins & 0x1)

    @mode.setter
    def mode(self, newmode: MuxMode):
        with self._ftdi_context() as ctx:
            # upper half: set all pins to output mode
            # lowest bit: enable or disable output for pin 0
            # (enable for host mode, disable for target mode)
            pins = 0xF0 | newmode.value
            ctx.set_bitmode(pins, Ftdi.BitMode.CBUS)
```

the reset appears to come down to one specific line in the `open_from_device` method:

https://github.com/eblot/pyftdi/blob/998583bdde59ec98d4dec67f141ec9992076188e/pyftdi/ftdi.py#L552

as a workaround, i've defined a subclass of `Ftdi` which overrides that method with that one specific line commented out. this is not ideal, because i'll potentially need to make manual changes any time i pull in a new version of PyFTDI, and because i'm not confident that your library is designed to work with a device initialized in an unknown state (although it does appear to work for my use-case).
